pkg/securitypolicy/securitypolicyenforcer_rego.go

```diff
@@ -992,92 +992,88 @@ func (policy *regoEnforcer) EnforceScratchUnmountPolicy(ctx context.Context, scr
 	return nil
 }
 
-func getUser(passwdPath string, filter func(user.User) bool) (user.User, error) {
-	users, err := user.ParsePasswdFileFilter(passwdPath, filter)
-	if err != nil {
-		return user.User{}, err
-	}
-	if len(users) != 1 {
-		return user.User{}, errors.Errorf("expected exactly 1 user matched '%d'", len(users))
-	}
-	return users[0], nil
-}
-
-func getGroup(groupPath string, filter func(user.Group) bool) (user.Group, error) {
-	groups, err := user.ParseGroupFileFilter(groupPath, filter)
-	if err != nil {
-		return user.Group{}, err
-	}
-	if len(groups) != 1 {
-		return user.Group{}, errors.Errorf("expected exactly 1 group matched '%d'", len(groups))
-	}
-	return groups[0], nil
-}
-
-func (policy *regoEnforcer) GetUserInfo(containerID string, process *oci.Process) (IDName, []IDName, string, error) {
-	rootPath := filepath.Join(guestpath.LCOWRootPrefixInUVM, containerID, guestpath.RootfsPath)
+func (policy *regoEnforcer) GetUserInfo(process *oci.Process, rootPath string) (
+	userIDName IDName,
+	groupIDNames []IDName,
+	umask string,
+	err error,
+) {
 	passwdPath := filepath.Join(rootPath, "/etc/passwd")
 	groupPath := filepath.Join(rootPath, "/etc/group")
 
 	if process == nil {
-		return IDName{}, nil, "", errors.New("spec.Process is nil")
+		err = errors.New("spec.Process is nil")
+		return
 	}
 
 	// this default value is used in the Linux kernel if no umask is specified
-	umask := "0022"
+	umask = "0022"
 	if process.User.Umask != nil {
 		umask = fmt.Sprintf("%04o", *process.User.Umask)
 	}
 
+	parsedUserStr := false
+	checkGroup := true
+
 	if process.User.Username != "" {
-		uid, gid, err := specGuest.ParseUserStr(rootPath, process.User.Username)
-		if err == nil {
-			userIDName := IDName{ID: strconv.FormatUint(uint64(uid), 10)}
-			groupIDName := IDName{ID: strconv.FormatUint(uint64(gid), 10)}
-			return userIDName, []IDName{groupIDName}, umask, nil
+		var usrInfo specGuest.ParseUserStrResult
+		usrInfo, err = specGuest.ParseUserStr(rootPath, process.User.Username)
+		if err != nil {
+			err = errors.Errorf("failed to parse user str %q: %v", process.User.Username, err)
+			return
 		}
-		log.G(context.Background()).WithError(err).Warn("failed to parse user str, fallback to lookup")
+		userIDName = IDName{ID: strconv.FormatUint(uint64(usrInfo.UID), 10), Name: usrInfo.Username}
+		groupIDNames = []IDName{
+			{ID: strconv.FormatUint(uint64(usrInfo.GID), 10), Name: usrInfo.Groupname},
+		}
+		parsedUserStr = true
 	}
 
-	// fallback UID/GID lookup
-	uid := process.User.UID
-	userIDName := IDName{ID: strconv.FormatUint(uint64(uid), 10), Name: ""}
-	if _, err := os.Stat(passwdPath); err == nil {
-		userInfo, err := getUser(passwdPath, func(user user.User) bool {
-			return uint32(user.Uid) == uid
-		})
+	if !parsedUserStr {
```

Comment:
Can this fallback code also be refactored? I think the `ParseUserStr` method already has same/similar code. Can we do something like :

```
if process.User.Username != "" {
    specGuest.ParseUserStr(rootPath, process.User.Username)
} else {
    specGuest.ParseUserStr(rotPath, fmt.Sprintf("%d:%d",process.User.UID, process.User.GID)
}
```
}